### PR TITLE
coll/gpu: Add GPU fallback paths and testing for reduce, allreduce

### PR DESCRIPTION
**Pull request by tarudoodi**
## Pull Request Description
Enables testing of MPI_Reduce and MPI_Allreduce with GPU buffers. 
- Adds fallback path in reduce and allreduce when the buffers are in GPU.
- Enables testing with GPU buffers in allred.c, allred2.c and reduce.c.  Other tests had user operations so we did not modify those.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by tarudoodi** [comment]
@raffenet Hi, this is the PR for the fallback paths. What would be the best way of testing it here? It passes all tests in our local testing.

**Comment by raffenet** [comment]
@tarudoodi I'll have to do some manual testing next week when I can get to it. Thanks for submitting!

**Comment by tarudoodi** [comment]
Pushed warning fixes.

**Comment by raffenet** [comment]
Initial comments:

1. Do we really need the fallback path in `MPI_[ALL]REDUCE_IMPL` functions? Presumably callers of those functions are either fallback calls by the device (buffers already replaced in MPIR) or other internal usage where we can assume no device buffers. Maybe I am missing some scenario?

2. The buffer replacement code looks almost (completely?) identical in both reduce and allreduce. If so, helper functions would be great to cut down the duplication.

File: src/mpi/coll/allreduce/allreduce.c [hunk 1]
```diff
@@ -235,6 +276,25 @@ int MPIR_Allreduce_impl(const void *sendbuf, void *recvbuf, int count, MPI_Datat
...
+     * and free the host recv buffer. */
+    if (recvbuf_in_device == 1) {
+        recvbuf = in_recvbuf;
+        MPIR_Typerep_copy(recvbuf, host_recvbuf, extent * count);
+        MPL_gpu_free_host(host_recvbuf);
+    }
+
+    goto fn_exit;
+
+  fallback:
```
Comment (raffenet) [inline src/mpi/coll/allreduce/allreduce.c:294]: this label is unused

File: test/mpi/maint/test_coll_algos.sh [hunk 2]
```diff
@@ -141,8 +141,8 @@ for algo_name in ${algo_names}; do
                 env="${testing_env} env=MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM=${algo_name}"
                 env="${env} env=MPIR_CVAR_IALLREDUCE_RECEXCH_KVAL=${kval}"
 
-                echo "allred 4 arg=100 ${env}" >> ${testlist_cvar}
-                echo "allred 7 ${env}" >> ${testlist_cvar}
+                echo "allred 4 arg=-count=100 arg=-evenmemtype=host arg=-oddmemtype=host ${env}" >> ${testlist_cvar}
```
Comment (raffenet) [inline test/mpi/maint/test_coll_algos.sh:144]: We made it so the MTest macros default to host memory type if not specified, so these extra arguments are not needed.

File: src/mpi/coll/allreduce/allreduce.c [hunk 3]
```diff
@@ -183,6 +183,47 @@ int MPIR_Allreduce_impl(const void *sendbuf, void *recvbuf, int count, MPI_Datat
                         MPI_Op op, MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPL_pointer_attr_t send_attr, recv_attr;
+    int sendbuf_in_device = 0, recvbuf_in_device = 0;
```
Comment (raffenet) [inline src/mpi/coll/allreduce/allreduce.c:187]: style nitpicks: I'd prefer `bool` types and `if (buf_in_device)` comparisons, i.e. no `== 1`.

**Review by raffenet — CHANGES_REQUESTED** [review]

**Comment by tarudoodi** [comment]
> Initial comments:
> 
> 1. Do we really need the fallback path in `MPI_[ALL]REDUCE_IMPL` functions? Presumably callers of those functions are either fallback calls by the device (buffers already replaced in MPIR) or other internal usage where we can assume no device buffers. Maybe I am missing some scenario?

In the `_impl` functions, we didn't see any use cases with device buffers as of now.  I assumed that someone could possibly call `MPIR_Allreduce_impl` with device buffers, so this will take care of it. If we assume what you're saying then we won't need them.

> 2. The buffer replacement code looks almost (completely?) identical in both reduce and allreduce. If so, helper functions would be great to cut down the duplication.

Yeah, the code is identical. The only reason I didn't create cleaner helper alloc/free functions is because they might need to duplicate the `extent` calculation. If I leave out the extent calculation it will save duplication of about 15 lines.  Do you have a preference?

File: test/mpi/maint/test_coll_algos.sh (hunk 2 above)
Comment (tarudoodi) [reply test/mpi/maint/test_coll_algos.sh:144]: Ok. I missed that part. I will look into it again and fix these.

**Comment by raffenet** [comment]
> > Initial comments:
> > 
> > 1. Do we really need the fallback path in `MPI_[ALL]REDUCE_IMPL` functions? Presumably callers of those functions are either fallback calls by the device (buffers already replaced in MPIR) or other internal usage where we can assume no device buffers. Maybe I am missing some scenario?
> 
> In the `_impl` functions, we didn't see any use cases with device buffers as of now. I assumed that someone could possibly call `MPIR_Allreduce_impl` with device buffers, so this will take care of it. If we assume what you're saying then we won't need them.

OK, let's drop the `_impl` changes for now.

> 
> > 1. The buffer replacement code looks almost (completely?) identical in both reduce and allreduce. If so, helper functions would be great to cut down the duplication.
> 
> Yeah, the code is identical. The only reason I didn't create cleaner helper alloc/free functions is because they might need to duplicate the `extent` calculation. If I leave out the extent calculation it will save duplication of about 15 lines. Do you have a preference?

This is probably not critical for the first pass. We can improve it later.

**Comment by tarudoodi** [comment]
> > > Initial comments:
> > > 
> > > 1. Do we really need the fallback path in `MPI_[ALL]REDUCE_IMPL` functions? Presumably callers of those functions are either fallback calls by the device (buffers already replaced in MPIR) or other internal usage where we can assume no device buffers. Maybe I am missing some scenario?
> > 
> > 
> > In the `_impl` functions, we didn't see any use cases with device buffers as of now. I assumed that someone could possibly call `MPIR_Allreduce_impl` with device buffers, so this will take care of it. If we assume what you're saying then we won't need them.
> 
> OK, let's drop the `_impl` changes for now.
> 
> > > 1. The buffer replacement code looks almost (completely?) identical in both reduce and allreduce. If so, helper functions would be great to cut down the duplication.
> > 
> > 
> > Yeah, the code is identical. The only reason I didn't create cleaner helper alloc/free functions is because they might need to duplicate the `extent` calculation. If I leave out the extent calculation it will save duplication of about 15 lines. Do you have a preference?
> 
> This is probably not critical for the first pass. We can improve it later.

Ok. I will update it with the changes tomorrow.

**Comment by raffenet** [comment]
BTW, I tested this manually with both ch4:ofi and ch4:ucx and the new tests look to work fine.

File: src/mpi/coll/allreduce/allreduce.c (hunk 1 above)
Comment (tarudoodi) [reply src/mpi/coll/allreduce/allreduce.c]: Fixed.

File: src/mpi/coll/allreduce/allreduce.c (hunk 3 above)
Comment (tarudoodi) [reply src/mpi/coll/allreduce/allreduce.c]: Updated.

File: test/mpi/maint/test_coll_algos.sh (hunk 2 above)
Comment (tarudoodi) [reply test/mpi/maint/test_coll_algos.sh]: > BTW, I tested this manually with both ch4:ofi and ch4:ucx and the new tests look to work fine.

That's great :)

**Comment by tarudoodi** [comment]
> BTW, I tested this manually with both ch4:ofi and ch4:ucx and the new tests look to work fine.

That's great :)

File: test/mpi/maint/generate_gpu_tests.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    # testlist.gpu
+    found=0
+    for dir in $dirs ; do
+        if test "$testdir" = "$dir" ; then
+            found=1
+        fi
+    done
+    if test "$found" = "0" ; then
+        dirs="$dirs $testdir"
+        printf "" >> ${builddir}/${testdir}/testlist.gpu
```
Comment (raffenet) [inline test/mpi/maint/generate_gpu_tests.sh:49]:
```suggestion
        printf "" > ${builddir}/${testdir}/testlist.gpu
```
I ran configure several times when doing manual tests, which resulted in a testlist.gpu file with many duplicate entries. Should we make sure to overwrite the file each time this test is run?

File: test/mpi/util/mtest_common.c [hunk 5]
```diff
@@ -315,7 +315,7 @@ void MTestAlloc(size_t size, mtest_mem_type_e type, void **hostbuf, void **devic
 #endif
 
     if (type == MTEST_MEM_TYPE__UNREGISTERED_HOST) {
-        *devicebuf = malloc(size);
+        *devicebuf = calloc(size, 1);
```
Comment (raffenet) [inline test/mpi/util/mtest_common.c:318]: This change should be split out. I'm not sure it's a good thing to always zero-init using this API. Maybe a flag could be added to the API, or a separate interface.

File: test/mpi/maint/generate_gpu_tests.sh (hunk 4 above)
Comment (tarudoodi) [reply test/mpi/maint/generate_gpu_tests.sh:49]: The `testlist.gpu` is first created by `gentests_dtp.sh` so my intent was to append to it. While rebasing today, I noticed that I had accidently removed the similar line of code from `gentests_dtp.sh`, which would make this line append over and over again. I think the current case should work but I will verify it again.

File: test/mpi/util/mtest_common.c (hunk 5 above)
Comment (tarudoodi) [reply test/mpi/util/mtest_common.c:318]: I can add a flag to the API and create a separate commit.

File: test/mpi/maint/generate_gpu_tests.sh (hunk 4 above)
Comment (raffenet) [reply test/mpi/maint/generate_gpu_tests.sh:49]: OK, great.

Comment (tarudoodi) [reply test/mpi/maint/generate_gpu_tests.sh:49]: I verified this and its working as expected.

File: test/mpi/util/mtest_common.c (hunk 5 above)
Comment (tarudoodi) [reply test/mpi/util/mtest_common.c]: Created a separate commit for this.

**Comment by raffenet** [comment]
Looks pretty good. I made some minor cleanups during review. Re-run the Jenkins tests. I'll do some manual GPU testing, as well.
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
So it occurs to me that we also need to modify `MPI_IREDUCE, MPI_IALLREDUCE, MPI_SCAN, MPI_ISCAN, MPI_EXSCAN, MPI_IEXSCAN` to complete GPU support for MPIR collectives. Now I'm more strongly for a helper function for doing the buffer manipulation. I'll think about this a little more and propose an API. @tarudoodi if you have something in the meantime, please share it.

**Comment by tarudoodi** [comment]
> So it occurs to me that we also need to modify `MPI_IREDUCE, MPI_IALLREDUCE, MPI_SCAN, MPI_ISCAN, MPI_EXSCAN, MPI_IEXSCAN` to complete GPU support for MPIR collectives. Now I'm more strongly for a helper function for doing the buffer manipulation. I'll think about this a little more and propose an API. @tarudoodi if you have something in the meantime, please share it.

I had a thought that maybe we have the first function(alloc) do all the calculation and let it set the values for extent, Xbuf_in_device, etc and let the second function use those values and only do free and the copy. It makes them a bit dependent but it fits the known use cases so far.

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
@tarudoodi here is my proposal for helpers. There are 2 main differences:
1. Use regular `MPL_malloc` to allocate the host buffers. Limits the amount of pinned memory used.
2. Use `MPIR_Localcopy` to copy to and from the device buffer. Avoids copying the gaps in non-contiguous types.

Let me know what you think.

**Comment by raffenet** [comment]
> @tarudoodi here is my proposal for helpers. There are 2 main differences:
> 
> 1. Use regular `MPL_malloc` to allocate the host buffers. Limits the amount of pinned memory used.
> 2. Use `MPIR_Localcopy` to copy to and from the device buffer. Avoids copying the gaps in non-contiguous types.
> 
> Let me know what you think.

The other difference is the datatype extent checks may be duplicated, but I'm not very concerned about it.

**Comment by tarudoodi** [comment]
> > @tarudoodi here is my proposal for helpers. There are 2 main differences:
> > 
> > 1. Use regular `MPL_malloc` to allocate the host buffers. Limits the amount of pinned memory used.
> > 2. Use `MPIR_Localcopy` to copy to and from the device buffer. Avoids copying the gaps in non-contiguous types.
> > 
> > Let me know what you think.
> 
> The other difference is the datatype extent checks may be duplicated, but I'm not very concerned about it.

@raffenet This looks good to me. I like taking benefit of `MPIR_Localcopy` for the non-contiguous datatypes. I added a patch that may help avoid duplication of extent calculation.

I only just noticed the failures in asan testing. Do you know what may be causing them? I will check it tomorrow.

**Comment by tarudoodi** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
> @raffenet This looks good to me. I like taking benefit of `MPIR_Localcopy` for the non-contiguous datatypes. I added a patch that may help avoid duplication of extent calculation.

Looks good to me.

> I only just noticed the failures in asan testing. Do you know what may be causing them? I will check it tomorrow.

Need to destroy the arglist in `allred`. I just pushed a fixup patch.

**Comment by tarudoodi** [comment]
Thanks for fixing the memory leak. I can merge all the fixup patches and update the PR accordingly today.

**Comment by raffenet** [comment]
> Thanks for fixing the memory leak. I can merge all the fixup patches and update the PR accordingly today.

Sounds good. I suggest adding the helper functions as a standalone commit, then adding them to all the collectives that need them. Let me know if you want any help.

**Comment by tarudoodi** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by raffenet — APPROVED** [review]

**Pull request merged**